smartsim/experiment.py

```diff
@@ -160,28 +168,45 @@ def __init__(
 
         self.exp_path = exp_path
 
-        self._launcher = launcher.lower()
-
-        if self._launcher == "auto":
-            self._launcher = detect_launcher()
-        if self._launcher == "cobalt":
-            raise SSUnsupportedError("Cobalt launcher is no longer supported.")
-
-        if launcher == "dragon":
-            self._set_dragon_server_path()
-
-        self._control = Controller(launcher=self._launcher)
+        # TODO: Remove this! The contoller is becoming obsolete
+        self._control = Controller(launcher="local")
+        self._dispatcher = settings_dispatcher
+        self._active_launchers: set[LauncherLike[t.Any]] = set()
```

Comment:
> we will perform a check using the active launchers and throw if they do not work together?

Sort of... The idea is that, with this new API, the user does not decide which launcher they want to use at `Experiment` construction time, and then is bound to that choice for the entirety of the lifetime of the object, but instead, the `Experiment` will set up multiple launchers on the fly and based on the types of jobs that a user passes to the experiment. For example:
```py
exp = Experiment("my-experiment", "/tmp/my/exp/path")  # <-- No launchers created here

slurm_settings = LaunchSettings("slurm")
app = Application(...)
slurm_job = Job(app, slurm_settings)
exp.start_jobs(slurm_job)  # <-- A shell launcher is created here and
                           #     added to the `_active_launchers` set

dragon_settings = LaunchSettings("dragon")
app_2 = Application(...)
dragon_job = Job(app_2, dragon_settings)
exp.start_jobs(dragon_job)  # <-- A dragon launcher is created here and 
                            #     added to the `_active_launchers` set
```
The active launchers set is maintianed so that the experiment does not need to set up a new launcher everytime they get a new job request of the same type
```py
slurm_settings_2 = LaunchSettings("slurm")
app_3 = Application(...)
another_slurm_job = Job(app_3, slurm_settings_2)
exp.start_jobs(another_slurm_job)  # <-- The previous shell launcher is used here,
                                   #     nothing new is constructed
```

Reply:
> Also, how can a workload have multiple Launcher? They would be launched on different machines?

They could be! If we want to start considering remote workflows, that is absolutely a possibility!!

For a more concrete example on a single machine, think about trying to launch a job using dragon and another job using raw slurm in SmartSim V0.7. Currently you cannot do this without two different `Experiment` instances. Versus with this design, the experiment can standup and use both launchers on the fly!

Reply:
> we will perform a check using the active launchers and throw if they do not work together?

This is also something we could do if the need came about, but so far I have been banking on the assumption that the launchers should be relatively independent of each other and there shouldn't be much (if any) cross talk between them which should hopefully limit incompatibility issues!

Reply:
> The active launchers set is maintianed so that the experiment does not need to set up a new launcher everytime they get a new job request of the same type

would you add this as a comment over active_launchers?